# ***Model Ensebling***

---
# **A. Voting**
---
# **1.Voting Calssifier**

# from sklearn.ensemble import VotingClassifier



In [ ]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import VotingClassifier
from sklearn.model_selection import train_test_split, GridSearchCV,StratifiedKFold
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score,log_loss
from sklearn.svm import SVC

In [ ]:
bcancer = pd.read_csv("/content/drive/MyDrive/KP_Module/Practical Machine Learning/Practise/Cases/Wisconsin/BreastCancer.csv",index_col=0)


In [ ]:
lbl = LabelEncoder()
X = bcancer.drop('Class',axis = 1)
y = pd.DataFrame(lbl.fit_transform(bcancer['Class']))


In [ ]:
X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=23,test_size=0.3,stratify=y)

svm = SVC(probability=True,random_state=23)
lr = LogisticRegression()
nb = GaussianNB()

voting = VotingClassifier(estimators=[("SVM",svm),("LR",lr),("NB",nb)],voting='soft')
voting.fit(X_train,y_train)

y_pred_prob = voting.predict_proba(X_test)[:,1]
print("Logloss : ",log_loss(y_test,y_pred_prob))
y_pred = voting.predict(X_test)
print("Accuraccy Score : ",accuracy_score(y_test,y_pred))

Logloss :  0.13776571787366462
Accuraccy Score :  0.9571428571428572


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:99: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)


In [ ]:
kfold = StratifiedKFold(n_splits=5,shuffle=True,random_state=23)
print(voting.get_params())
params = {'SVM__gamma':['scale','auto'],
          'SVM__C':np.linspace(0.001,5,5),
          'LR__penalty':['l1','l2','elastic',None],
          'NB__var_smoothing':np.linspace(0.0001,0.999,20)}
gcv = GridSearchCV(voting,param_grid=params,cv=kfold,scoring='neg_log_loss')

gcv.fit(X,y)
print("Best Parameters : ",gcv.best_params_)
print("Best Score : ",gcv.best_score_)

{'estimators': [('SVM', SVC(probability=True, random_state=23)), ('LR', LogisticRegression()), ('NB', GaussianNB())], 'flatten_transform': True, 'n_jobs': None, 'verbose': False, 'voting': 'soft', 'weights': None, 'SVM': SVC(probability=True, random_state=23), 'LR': LogisticRegression(), 'NB': GaussianNB(), 'SVM__C': 1.0, 'SVM__break_ties': False, 'SVM__cache_size': 200, 'SVM__class_weight': None, 'SVM__coef0': 0.0, 'SVM__decision_function_shape': 'ovr', 'SVM__degree': 3, 'SVM__gamma': 'scale', 'SVM__kernel': 'rbf', 'SVM__max_iter': -1, 'SVM__probability': True, 'SVM__random_state': 23, 'SVM__shrinking': True, 'SVM__tol': 0.001, 'SVM__verbose': False, 'LR__C': 1.0, 'LR__class_weight': None, 'LR__dual': False, 'LR__fit_intercept': True, 'LR__intercept_scaling': 1, 'LR__l1_ratio': None, 'LR__max_iter': 100, 'LR__multi_class': 'auto', 'LR__n_jobs': None, 'LR__penalty': 'l2', 'LR__random_state': None, 'LR__solver': 'lbfgs', 'LR__tol': 0.0001, 'LR__verbose': 0, 'LR__warm_start': False, 'NB_

Streaming output truncated to the last 5000 lines.
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:99: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:99: DataConversionWarning:

Best Parameters :  {'LR__penalty': None, 'NB__var_smoothing': 0.31554210526315785, 'SVM__C': 2.5004999999999997, 'SVM__gamma': 'auto'}
Best Score :  -0.0945575859740283


# Image Segmentation Dataset

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/KP_Module/Practical Machine Learning/Practise/Cases/Image Segmentation/Image_Segmention.csv")

lbl = LabelEncoder()
X = data.drop('Class',axis = 1)
y = pd.DataFrame(lbl.fit_transform(data['Class']))

X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=23,test_size=0.3,stratify=y)

svm = SVC(probability=True,random_state=23)
lr = LogisticRegression()
nb = GaussianNB()

#model Assembling
voting = VotingClassifier(estimators=[("SVM",svm),("LR",lr),("NB",nb)],voting='soft')
voting.fit(X_train,y_train)

y_pred_prob = voting.predict_proba(X_test)#[:,1]
print("Logloss : ",log_loss(y_test,y_pred_prob))
y_pred = voting.predict(X_test)
print("Accuraccy Score : ",accuracy_score(y_test,y_pred))

#gridsearchcv
kfold = StratifiedKFold(n_splits=5,shuffle=True,random_state=23)
print(voting.get_params())
params = {'SVM__gamma':['scale','auto'],
          'SVM__C':np.linspace(0.001,5,5),
          'LR__penalty':['l1','l2','elastic',None],
          "LR__multi_class":['ovr','multinomal'],
          'NB__var_smoothing':np.linspace(0.0001,0.999,20)}
gcv = GridSearchCV(voting,param_grid=params,cv=kfold,scoring='neg_log_loss')

gcv.fit(X,y)
print("Best Parameters : ",gcv.best_params_)
print("Best Score : ",gcv.best_score_)

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:99: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear

Logloss :  0.4045324747689428
Accuraccy Score :  0.8888888888888888
{'estimators': [('SVM', SVC(probability=True, random_state=23)), ('LR', LogisticRegression()), ('NB', GaussianNB())], 'flatten_transform': True, 'n_jobs': None, 'verbose': False, 'voting': 'soft', 'weights': None, 'SVM': SVC(probability=True, random_state=23), 'LR': LogisticRegression(), 'NB': GaussianNB(), 'SVM__C': 1.0, 'SVM__break_ties': False, 'SVM__cache_size': 200, 'SVM__class_weight': None, 'SVM__coef0': 0.0, 'SVM__decision_function_shape': 'ovr', 'SVM__degree': 3, 'SVM__gamma': 'scale', 'SVM__kernel': 'rbf', 'SVM__max_iter': -1, 'SVM__probability': True, 'SVM__random_state': 23, 'SVM__shrinking': True, 'SVM__tol': 0.001, 'SVM__verbose': False, 'LR__C': 1.0, 'LR__class_weight': None, 'LR__dual': False, 'LR__fit_intercept': True, 'LR__intercept_scaling': 1, 'LR__l1_ratio': None, 'LR__max_iter': 100, 'LR__multi_class': 'auto', 'LR__n_jobs': None, 'LR__penalty': 'l2', 'LR__random_state': None, 'LR__solver': 'lbfgs'

Streaming output truncated to the last 5000 lines.
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:99: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change 

Best Parameters :  {'LR__penalty': None, 'NB__var_smoothing': 0.0001, 'SVM__C': 5.0, 'SVM__gamma': 'scale'}
Best Score :  -0.47018783553377486


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
2000 fits failed out of a total of 4000.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1000 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_voting.py", line 346, in fit
    return super().fit(X, transformed_y, sample_weight)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_voting.py", line 81, in fit
    self.estimators_ = Parallel(n_jobs=self.n_jobs)(
  File "/usr

# in voting Classifier there is a weight parameter that needs to be fine tune

---

**if we calculate the indvidual scores of each of the estimators(using gcv only), then the best performer should get best weight and worst performergets low weight**

**In Above Case LR has best score followed by SVM and NB shows poor performancwe so we assign weigths to lr as 0.5 , svm as 0.4 and nb as 0.1, Note these wieghts are completely random**


In [ ]:
X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=23,test_size=0.3,stratify=y)

svm = SVC(probability=True,random_state=23)
lr = LogisticRegression()
nb = GaussianNB()

voting = VotingClassifier(estimators=[("SVM",svm),("LR",lr),("NB",nb)],voting='soft',weights=[0.5,0.4,0.1])
voting.fit(X_train,y_train)

y_pred_prob = voting.predict_proba(X_test)#[:,1]
print("Logloss : ",log_loss(y_test,y_pred_prob))
y_pred = voting.predict(X_test)
print("Accuraccy Score : ",accuracy_score(y_test,y_pred))

Logloss :  0.4466073969843571
Accuraccy Score :  0.9047619047619048


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:99: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear

In [ ]:
#gridsearchcv
kfold = StratifiedKFold(n_splits=5,shuffle=True,random_state=23)
print(voting.get_params())
params = {'SVM__gamma':['scale','auto'],
          'SVM__C':np.linspace(0.001,5,5),
          'LR__penalty':['l1','l2','elastic',None],
          "LR__multi_class":['ovr','multinomal'],
          'NB__var_smoothing':np.linspace(0.0001,0.999,20)}
gcv = GridSearchCV(voting,param_grid=params,cv=kfold,scoring='neg_log_loss')

gcv.fit(X,y)
print("Best Parameters : ",gcv.best_params_)
print("Best Score : ",gcv.best_score_)

{'estimators': [('SVM', SVC(probability=True, random_state=23)), ('LR', LogisticRegression()), ('NB', GaussianNB())], 'flatten_transform': True, 'n_jobs': None, 'verbose': False, 'voting': 'soft', 'weights': [0.5, 0.4, 0.1], 'SVM': SVC(probability=True, random_state=23), 'LR': LogisticRegression(), 'NB': GaussianNB(), 'SVM__C': 1.0, 'SVM__break_ties': False, 'SVM__cache_size': 200, 'SVM__class_weight': None, 'SVM__coef0': 0.0, 'SVM__decision_function_shape': 'ovr', 'SVM__degree': 3, 'SVM__gamma': 'scale', 'SVM__kernel': 'rbf', 'SVM__max_iter': -1, 'SVM__probability': True, 'SVM__random_state': 23, 'SVM__shrinking': True, 'SVM__tol': 0.001, 'SVM__verbose': False, 'LR__C': 1.0, 'LR__class_weight': None, 'LR__dual': False, 'LR__fit_intercept': True, 'LR__intercept_scaling': 1, 'LR__l1_ratio': None, 'LR__max_iter': 100, 'LR__multi_class': 'auto', 'LR__n_jobs': None, 'LR__penalty': 'l2', 'LR__random_state': None, 'LR__solver': 'lbfgs', 'LR__tol': 0.0001, 'LR__verbose': 0, 'LR__warm_start': 

Streaming output truncated to the last 5000 lines.
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:99: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
ST

Best Parameters :  {'LR__penalty': None, 'NB__var_smoothing': 0.0001, 'SVM__C': 5.0, 'SVM__gamma': 'scale'}
Best Score :  -0.46014105440437747


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
2000 fits failed out of a total of 4000.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1000 fits failed with the following error:
T

#Result of using weights parameter in votingClassifier
here we can see after taking the weights parameter the results are improved from -0.47018 to -0.46014

In [ ]:
# Do the same thing by removing the GaussianNB by DecisionTree
# this take To much time so dont Implement
# for use randomzedsearchcv

# ***2. VotingRegressor***
---
from sklearn.ensemble import VotingRegressor

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import RandomizedSearchCV,train_test_split,KFold
from sklearn.ensemble import VotingRegressor
from sklearn.linear_model import ElasticNet
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import StandardScaler,MinMaxScaler
from sklearn.pipeline import Pipeline

In [ ]:
concrete = pd.read_csv("/content/drive/MyDrive/KP_Module/Practical Machine Learning/Practise/Cases/Concrete Strength/Concrete_Data.csv")

In [ ]:
X = concrete.drop('Strength',axis = 1)
y = concrete['Strength']

In [ ]:
kfold = KFold(n_splits=5,shuffle=True,random_state=23)

elastic = ElasticNet()
std_scl = StandardScaler()
min_max = MinMaxScaler()
knn = KNeighborsRegressor()
dtr = DecisionTreeRegressor()

pipe = Pipeline([("SCL",std_scl),("KNN",knn)])

voting = VotingRegressor(estimators=[('ELASTIC',elastic),('PIPE',pipe),('DTR',dtr)])

params = {"ELASTIC__alpha":np.linspace(0.001,5,5),
          'ELASTIC__l1_ratio':np.linspace(0,1,5),
          'PIPE__KNN__n_neighbors':np.arange(1,10),
          'PIPE__SCL':[std_scl,min_max],
          'DTR__max_depth':[None,3,5],
          'DTR__min_samples_split':[2,5,10],
          'DTR__min_samples_leaf':[1,5,10]}


rgcv = RandomizedSearchCV(voting,param_distributions=params,cv=kfold,scoring='r2')

rgcv.fit(X,y)

print("Best Parameters : ",rgcv.best_params_)
print("Best Score : ",rgcv.best_score_)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.535e+04, tolerance: 2.258e+01 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge/RidgeCV instead.
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.331e+04, tolerance: 2.326e+01 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_mode

Best Parameters :  {'PIPE__SCL': MinMaxScaler(), 'PIPE__KNN__n_neighbors': 4, 'ELASTIC__l1_ratio': 0.5, 'ELASTIC__alpha': 0.001, 'DTR__min_samples_split': 5, 'DTR__min_samples_leaf': 1, 'DTR__max_depth': None}
Best Score :  0.8244883044278716


In [ ]:
from sklearn.model_selection import GridSearchCV


rgcv = GridSearchCV(voting,param_grid=params,cv=kfold,scoring='r2')

rgcv.fit(X,y)

print("Best Parameters : ",rgcv.best_params_)
print("Best Score : ",rgcv.best_score_)

Streaming output truncated to the last 5000 lines.
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.535e+04, tolerance: 2.258e+01 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge/RidgeCV instead.
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.331e+04, tolerance: 2.326e+01 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using on

Best Parameters :  {'DTR__max_depth': None, 'DTR__min_samples_leaf': 1, 'DTR__min_samples_split': 2, 'ELASTIC__alpha': 2.5004999999999997, 'ELASTIC__l1_ratio': 1.0, 'PIPE__KNN__n_neighbors': 1, 'PIPE__SCL': StandardScaler()}
Best Score :  0.8513578415741225


**Here we acn use Gridsearchcv but it takes too much time Here we havn't use weight parameter in votingregressor so to use that we have to check the individual performance of all the models**

**here R2 score for ElasticNet is 0.6006, for KNN-regressor it is 0.715 and for Decision tree regressor it is 0.847
so we are giving the more weightage to the DTR than KNN_R and less weightage to the ElasticNest**

**here we are taking weights equal to their corresponding R2 score
therefor we are using wights = [0.6006,0.715,0.847]**

In [ ]:
kfold = KFold(n_splits=5,shuffle=True,random_state=23)

elastic = ElasticNet()
std_scl = StandardScaler()
min_max = MinMaxScaler()
knn = KNeighborsRegressor()
dtr = DecisionTreeRegressor()

pipe = Pipeline([("SCL",std_scl),("KNN",knn)])

voting = VotingRegressor(estimators=[('ELASTIC',elastic),('PIPE',pipe),('DTR',dtr)],weights=[0.6006,0.715,0.847])

params = {"ELASTIC__alpha":np.linspace(0.001,5,5),
          'ELASTIC__l1_ratio':np.linspace(0,1,5),
          'PIPE__KNN__n_neighbors':np.arange(1,10),
          'PIPE__SCL':[std_scl,min_max],
          'DTR__max_depth':[None,3,5],
          'DTR__min_samples_split':[2,5,10],
          'DTR__min_samples_leaf':[1,5,10]}


rgcv = RandomizedSearchCV(voting,param_distributions=params,cv=kfold,scoring='r2')

rgcv.fit(X,y)

print("Best Parameters : ",rgcv.best_params_)
print("Best Score : ",rgcv.best_score_)

Best Parameters :  {'PIPE__SCL': StandardScaler(), 'PIPE__KNN__n_neighbors': 1, 'ELASTIC__l1_ratio': 0.25, 'ELASTIC__alpha': 1.2507499999999998, 'DTR__min_samples_split': 5, 'DTR__min_samples_leaf': 1, 'DTR__max_depth': None}
Best Score :  0.8580713131499701


# Kaggle Dataset

---

# Crab Age Prediction

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import RandomizedSearchCV,train_test_split,KFold
from sklearn.ensemble import VotingRegressor
from sklearn.linear_model import ElasticNet
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import StandardScaler,MinMaxScaler
from sklearn.pipeline import Pipeline

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/KP_Module/Practical Machine Learning/Practise/Cases/CRAB_AGE/train.csv",index_col=0)
data1=pd.get_dummies(data,drop_first=True)
X= data1.drop("Age",axis=1)
y = data.Age

In [ ]:
kfold = KFold(n_splits=5,shuffle=True,random_state=23)

elastic = ElasticNet()

knn = KNeighborsRegressor()
dtr = DecisionTreeRegressor()



voting = VotingRegressor(estimators=[('ELASTIC',elastic),('KNN',knn),('DTR',dtr)])#,weights=[0.6006,0.715,0.847]

params = {"ELASTIC__alpha":np.linspace(0.001,5,5),
          'ELASTIC__l1_ratio':np.linspace(0,1,5),
          'KNN__n_neighbors':np.arange(1,10),
          'DTR__max_depth':[None,3,5],
          'DTR__min_samples_split':[2,5,10],
          'DTR__min_samples_leaf':[1,5,10]}


rgcv = RandomizedSearchCV(voting,param_distributions=params,cv=kfold,scoring='r2')

rgcv.fit(X,y)

print("Best Parameters : ",rgcv.best_params_)
print("Best Score : ",rgcv.best_score_)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.740e+05, tolerance: 5.983e+01 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge/RidgeCV instead.
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.745e+05, tolerance: 5.992e+01 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_mode

Best Parameters :  {'KNN__n_neighbors': 4, 'ELASTIC__l1_ratio': 1.0, 'ELASTIC__alpha': 0.001, 'DTR__min_samples_split': 10, 'DTR__min_samples_leaf': 10, 'DTR__max_depth': 5}
Best Score :  0.5636392008890436


In [ ]:
#inferencing
test_data = pd.read_csv("/content/drive/MyDrive/KP_Module/Practical Machine Learning/Practise/Cases/CRAB_AGE/test.csv",index_col=0)
test_data = pd.get_dummies(test_data,drop_first=True)
best_model = rgcv.best_estimator_

y_pred = best_model.predict(test_data)

submission={'id':np.arange(74051,123419),'Age':y_pred}
submission=pd.DataFrame(submission)
submission.to_csv("/content/drive/MyDrive/KP_Module/Practical Machine Learning/Practise/Cases/CRAB_AGE/submission.csv",index=False)